File: applications/rtkdigisensgeometry/rtkdigisensgeometry.cxx
/*=========================================================================
 *
 *  Copyright RTK Consortium
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0.txt
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 *=========================================================================*/

#include "rtkdigisensgeometry_ggo.h"
#include "rtkGgoFunctions.h"

#include "rtkThreeDCircularProjectionGeometryXMLFile.h"
#include "rtkDigisensGeometryReader.h"

int main(int argc, char * argv[])
{
  GGO(rtkdigisensgeometry, args_info);

  // RTK geometry object
  typedef rtk::ThreeDCircularProjectionGeometry GeometryType;
  GeometryType::Pointer geometry = GeometryType::New();

  // Create geometry reader
  rtk::DigisensGeometryReader::Pointer reader = rtk::DigisensGeometryReader::New();
  reader->SetXMLFileName(args_info.xml_file_arg);
  TRY_AND_EXIT_ON_ITK_EXCEPTION( reader->UpdateOutputData() )

  // Write
  rtk::ThreeDCircularProjectionGeometryXMLFileWriter::Pointer xmlWriter =
    rtk::ThreeDCircularProjectionGeometryXMLFileWriter::New();
  xmlWriter->SetFilename(args_info.output_arg);
  xmlWriter->SetObject(&(*geometry) );
  TRY_AND_EXIT_ON_ITK_EXCEPTION( xmlWriter->WriteFile() )

  return EXIT_SUCCESS;
}

File: applications/rtkdigisensgeometry/rtkdigisensgeometry.py
import argparse
from itk import RTK as rtk

def build_parser():
    parser = argparse.ArgumentParser(
        description="Read Digisens geometry XML and write RTK geometry XML file.",
        formatter_class=argparse.ArgumentDefaultsHelpFormatter,
    )
    parser.add_argument(
        "--xml_file", "-x", required=True, help="Input Digisens XML calibration file"
    )
    parser.add_argument(
        "--output", "-o", required=True, help="Output geometry XML file"
    )
    parser.add_argument(
        "--verbose", "-v", action="store_true", help="Enable verbose output"
    )
    return parser

def process(args_info: argparse.Namespace):
    # Create geometry reader
    reader = rtk.DigisensGeometryReader.New()
    reader.SetXMLFileName(args_info.xml_file)
    if args_info.verbose:
        print(f"Reading Digisens XML: {args_info.xml_file}")
    reader.UpdateOutputData()

    # Write geometry
    if args_info.verbose:
        print(f"Writing geometry to {args_info.output}")
    rtk.write_geometry(reader.GetGeometry(), args_info.output)
    if args_info.verbose:
        print("Done.")

def main(argv=None):
    parser = build_parser()
    args_info = parser.parse_args(argv)
    process(args_info)

if __name__ == "__main__":
    main()